### PR TITLE
Questions as pattern rather than form

### PR DESCRIPTION
**Pull request by tmbo**
**Proposed changes**:
- added pattern for question
- allow using context variables in patterns 
- coerce string-boolean values of the llm command generator to proper booleans
- only include the slots than can be filled from the current position in the dialogue (e.g. slots where "skip_if_filled" is set to false can not be filled in advance of the question)
- added continue flow step to be able to resume at any point in a flow -> step has a "flexible" next, pointing to the step that should be resumed.
- renamed `skip_if_filled` to `ask_before_filling`

Fixes https://rasahq.atlassian.net/browse/ATO-1417
Fixes https://rasahq.atlassian.net/browse/ENG-504

### CODE REVIEW

**Comment by tmbo** [comment]
the code needs a little bit more cleanup, but would like feedback on the approach first and if you feel this is a good solution.

**Comment by tmbo** [comment]
I'd also go ahead and remove the "cleanup_step" again, because with the continuation node we actually don't need it and can use that instead.

File: rasa/cdu/command_processor.py [hunk 1]
```diff
@@ -358,8 +373,31 @@ def clean_up_commands(
             structlogger.debug(
                 "command_executor.skip_command.slot_already_set", command=command
             )
+        elif isinstance(command, SetSlotCommand) and command.name not in slots_so_far:
+            all_questions = [
+                s for f in all_flows.underlying_flows for s in f.get_question_steps()
```
Comment (radovanZRasa) [inline rasa/cdu/command_processor.py:378]: Sorry for para-trooping like this:
I am sure you are already familiar with this and I know that timing is tight, but we should avoid usage of one letter variable. It makes code harder to maintain and read for future developers.

Comment (tmbo) [reply rasa/cdu/command_processor.py:378]: weeeeel, i do disagree for for-comprehensions as you usually either need to compromise and split up the comprehension across multiple lines or keep it in a single line with a shorter name.

in this case though, fully agree esp. due to the "double comprehension".

File: rasa/cdu/command_processor.py [hunk 2]
```diff
@@ -358,8 +373,31 @@ def clean_up_commands(
...
+            ]
+            use_slot_fill = True
+            for q in all_questions:
+                if q.question == command.name and not q.skip_if_filled:
+                    use_slot_fill = False
```
Comment (twerkmeister) [inline rasa/cdu/command_processor.py:383]: this could potentially break if some generic slot is used with `skip_if_filled: True` in one flow and with `skip_if_filled: False` in another, but not sure anything can be done about it.

File: rasa/shared/core/flows/flow.py [hunk 3]
```diff
@@ -526,6 +529,34 @@ def __init__(self) -> None:
...
+            id=CONTINUE_STEP_PREFIX + next,
+            description=None,
+            metadata={},
+            # The end step links to itself. This is needed to make sure that
+            # this allows us to end a flow by setting the active step of a flow
+            # to the end step.
+            # Since the side effects of a node are executed on the transition
+            # to the next node, we need this link to run the END logic.
+            # Otherwise, setting a flow to its end step would not execute the
+            # side effects of the end step.
```
Comment (twerkmeister) [inline rasa/shared/core/flows/flow.py:551]: seems this was copied and pasted

**Review by twerkmeister — COMMENTED** [review]
Some comments for now ~ definitely changes things a bit. I think I need to run it to gain full confidence

File: rasa/cdu/command_generator/llm_command_generator.py [hunk 4]
```diff
@@ -248,10 +273,39 @@ def create_template_inputs(
...
+        tracker: DialogueStateTracker,
+        current_step: Optional[FlowStep],
+    ) -> bool:
         slot = tracker.slots.get(q.question)
         if slot is None:
             return False
+        # TODO: maybe we need to check for non None default value?
+        return (
+            q.skip_if_filled
+            or slot.value is not None
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/cdu/command_generator/llm_command_generator.py:287]: mhm yes we might need to work more on this logic. the "is value none" check is a heuristic to see if the question was asked without checking where we are in a flow and needing to calculate graph paths.

I agree that jumping back needs more thinking: maybe we should reset all the slots that have skip_if_filled set if we jump back "behind" them (assuming there is no loop in the flow)

File: rasa/cdu/command_processor.py (hunk 2 above)
Comment (tmbo) [reply rasa/cdu/command_processor.py:383]: mhm agreed but don't think we can resolve that can we

File: rasa/cdu/command_generator/llm_command_generator.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ancalita) [reply rasa/cdu/command_generator/llm_command_generator.py:306]: What about also checking that slot.value is not the same as `slot.initial_value` if that's set?

File: rasa/cdu/command_generator/llm_command_generator.py [hunk 5]
```diff
@@ -273,30 +321,43 @@ def allowed_values_for_slot(self, slot: Slot) -> Optional[str]:
...
+
     def render_template(
         self, message: Message, tracker: DialogueStateTracker, flows: FlowsList
     ) -> str:
         flows_without_patterns = FlowsList(
             [f for f in flows.underlying_flows if not f.is_handling_pattern()]
         )
         flow_stack = FlowStack.from_tracker(tracker)
-        top_flow = flow_stack.top_flow(flows) if flow_stack is not None else None
+        top_relevant_frame = flow_stack.top(ignore=FLOW_PATTERN_ASK_QUESTION)
```
Comment (ancalita) [inline rasa/cdu/command_generator/llm_command_generator.py:339]: Not sure I understand what the `ignore=` argument does?

File: rasa/cdu/command_processor.py [hunk 6]
```diff
@@ -155,17 +157,21 @@ def execute_commands(
...
-            if current_top_flow.id != FLOW_PATTERN_CORRECTION_ID:
+            if top_non_question_frame.flow_id != FLOW_PATTERN_CORRECTION_ID:
                 flow_stack.push(correction_frame)
             else:
                 # wrap up the previous correction flow
-                flow_stack.frames[-1].step_id = END_STEP
+                for i, frame in enumerate(reversed(flow_stack.frames)):
+                    frame.step_id = END_STEP
+                    if frame.frame_id == top_non_question_frame.frame_id:
+                        break
```
Comment (ancalita) [inline rasa/cdu/command_processor.py:167]: Is it correct to say that before we break at `top_non_question_frame`, we are adding an END_STEP to question frame? What's the implication of that?

File: rasa/cdu/command_processor.py [hunk 7]
```diff
@@ -358,8 +373,29 @@ def clean_up_commands(
...
+                step.question != command.name or step.skip_if_filled
+                for flow in all_flows.underlying_flows
+                for step in flow.get_question_steps()
+            )
+
+            if use_slot_fill:
+                clean_commands.append(command)
+            else:
+                structlogger.debug(
+                    "command_executor.skip_command.slot_not_asked_for", command=command
```
Comment (ancalita) [inline rasa/cdu/command_processor.py:387]: Shouldn't these branches be in reverse? Doesn't `step.skip_if_filled` is False and `step.question == command_name` mean that we should still fill the slot?

File: rasa/cdu/flow_stack.py [hunk 8]
```diff
@@ -101,27 +101,30 @@ def current_context(self) -> Dict[Text, Any]:
 
         return self.frames[-1].context or {}
 
-    def top(self) -> Optional[FlowStackFrame]:
+    def top(self, ignore: Optional[str] = None) -> Optional[FlowStackFrame]:
         """Returns the topmost frame from the stack.
```
Comment (ancalita) [inline rasa/cdu/flow_stack.py:105]: Would be great to add info about `ignore` arg in the docstring 🙏🏻

File: rasa/core/actions/flows.py [hunk 9]
```diff
@@ -156,10 +157,22 @@ async def run(
         corrected_slots = context.get("corrected_slots", {})
         reset_point = context.get("corrected_reset_point", {})
```
Comment (ancalita) [inline rasa/core/actions/flows.py:158]: What is the reset point in this context? This will help me understand why we break the below loop in L160 when the reset point id matches the flow id.

File: rasa/core/policies/default_flows.yml [hunk 10]
```diff
@@ -98,3 +98,23 @@ flows:
         next: "1"
       - id: "1"
         action: utter_clarification_options_rasa
+
+  pattern_ask_question:
+    description: flow used to fill a slot
+    steps:
+      - id: "start"
+        action: action_extract_slots
```
Comment (ancalita) [inline rasa/core/policies/default_flows.yml:106]: I'm assuming this allows the user to continue using existing custom extraction and validation actions from `rasa-sdk`? While the sdk `ValidationAction`still makes sense, how do we adapt away from usages of `FormValidationAction` (used only within form context)?

File: rasa/shared/core/flows/flow.py [hunk 11]
```diff
@@ -526,6 +529,34 @@ def __init__(self) -> None:
...
 
+CONTINUE_STEP_PREFIX = "__next__"
+
+
+@dataclass
+class ContinueFlowStep(InternalFlowStep):
+    """Represents the configuration of an end to a flow."""
+
+    def __init__(self, next: str) -> None:
+        """Initializes a start flow step."""
```
Comment (ancalita) [inline rasa/shared/core/flows/flow.py:540]: Docstring needs amending?

File: rasa/core/policies/flow_policy.py [hunk 12]
```diff
@@ -614,94 +559,97 @@ def _run_step(
...
-                return ActionPrediction(None, 0.0)
+            structlogger.debug("flow.step.run.question")
+
+            context = self.flow_stack.current_context().copy()
+            context["question"] = step.question
+
+            self.flow_stack.push(
+                FlowStackFrame(
+                    flow_id=FLOW_PATTERN_ASK_QUESTION,
+                    frame_type=StackFrameType.REMARK,
```
Comment (ancalita) [inline rasa/core/policies/flow_policy.py:570]: What is a remark stack frame type?

File: rasa/core/policies/flow_policy.py [hunk 13]
```diff
@@ -614,94 +559,97 @@ def _run_step(
...
+                    context=context,
+                )
+            )
 
-            question_action = ActionPrediction("question_" + step.question, 1.0)
-            if slot_value != initial_value:
-                question_action.events = [SlotSet(step.question, initial_value)]
-            return question_action
+            return ActionPrediction(None, 0.0)
```
Comment (ancalita) [inline rasa/core/policies/flow_policy.py:575]: More for my knowledge, an action prediction with these parameters allows the assistant to continue making predictions? 🤔

File: rasa/cdu/command_processor.py (hunk 1 above)
Comment (radovanZRasa) [reply rasa/cdu/command_processor.py]: `s` easily becomes `step` or `step_item` and `f` becomes `flow` or `flow_item`.
I believe that either comprehension should be a bot easier to read as well.

**Review by ancalita — COMMENTED** [review]
Thanks for working on this @tmbo 🙏🏻  I left a number of enquiries while trying to make sense of all code changes.

File: rasa/cdu/command_generator/llm_command_generator.py (hunk 5 above)
Comment (ancalita) [reply rasa/cdu/command_generator/llm_command_generator.py:339]: Discovered the reason further along the code 😄

File: rasa/cdu/command_generator/llm_command_generator.py [hunk 14]
```diff
@@ -177,8 +178,40 @@ def clean_extracted_value(value: str) -> str:
...
+
+        slot = tracker.slots[slot_name]
+        if isinstance(slot, BooleanSlot):
+            return bool_from_any(value)
+        elif isinstance(slot, FloatSlot):
+            try:
+                return float(slot)
+            except ValueError:
+                return None
+        else:
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/cdu/command_generator/llm_command_generator.py:208]: @twerkmeister that is actually a good question, do you feel the LLM is predicting a consistent enough format here that allows us to try to parse it?

File: rasa/cdu/command_processor.py (hunk 6 above)
Comment (tmbo) [reply rasa/cdu/command_processor.py:167]: we are not really adding a step to the question flow but rather we are setting the stack frame to the end step of that flow. so the step is already there but we "progress" that flow to its end step. 

tl;dr. we remove the question frame from the stack, but in a graceful way where the flow has a chance to wrap up in its end step.

File: rasa/cdu/command_processor.py (hunk 7 above)
Comment (tmbo) [reply rasa/cdu/command_processor.py:387]: yes and no 😂 so the branches are right but your hint that `use_slot_fill` wasn't set properly is a good one, there were two cases that wasn't properly covering :+1:

File: rasa/core/actions/flows.py (hunk 9 above)
Comment (tmbo) [reply rasa/core/actions/flows.py:158]: when a user corrects a the answer to a prior question, we jump back in the flow to that question to make sure, we follow the right branches afterwards -> a different answer to that question might lead us down a different path. the reset point is the step we need to jump too. (shouldn'T be called point though, will rename)

File: rasa/core/policies/default_flows.yml (hunk 10 above)
Comment (tmbo) [reply rasa/core/policies/default_flows.yml:106]: mhm good question, we probably should create a cleaned up validation action specifically for flows

File: rasa/core/policies/flow_policy.py (hunk 13 above)
Comment (tmbo) [reply rasa/core/policies/flow_policy.py]: yes indeed, but improved the code to make that more explicit.

File: rasa/cdu/command_generator/llm_command_generator.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/cdu/command_generator/llm_command_generator.py:205]: we won't tackle as part of this pr, but created a separate issue for this https://rasahq.atlassian.net/browse/ENG-513

**Review by twerkmeister — COMMENTED** [review]
more comments

File: rasa/cdu/command_generator/llm_command_generator.py [hunk 15]
```diff
@@ -198,13 +228,12 @@ def parse_commands(cls, actions: Optional[str]) -> List[Command]:
         for action in actions.strip().splitlines():
             if m := slot_set_re.search(action):
                 slot_name = m.group(1).strip()
-                slot_value: Optional[str] = cls.clean_extracted_value(m.group(2))
+                slot_value: Any = cls.clean_extracted_value(m.group(2))
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/cdu/command_generator/llm_command_generator.py:231]: it does need `any` because it is mutable and the value gets overwritten with the coerced value

File: rasa/core/policies/default_flows.yml [hunk 16]
```diff
@@ -51,11 +51,17 @@ flows:
     description: Confirm a correction of a slot value.
 
     steps:
+      - id: "check_if_reset_only"
+        next:
+          - if: context.is_reset_only
+            then: "done"
+          - else: "correct_slots"
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/core/policies/default_flows.yml:58]: only fixed that in the example, glad you caught it - didn't turn up as an issue because the demo overwrites it...

File: rasa/shared/core/flows/flow.py [hunk 17]
```diff
@@ -506,45 +506,42 @@ class EndFlowStep(InternalFlowStep):
...
             metadata={},
-            # The end step links to itself. This is needed to make sure that
-            # this allows us to end a flow by setting the active step of a flow
-            # to the end step.
-            # Since the side effects of a node are executed on the transition
-            # to the next node, we need this link to run the END logic.
-            # Otherwise, setting a flow to its end step would not execute the
-            # side effects of the end step.
-            next=FlowLinks(links=[StaticFlowLink(target=CLEANUP_STEP)]),
+            next=FlowLinks(links=[]),
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/shared/core/flows/flow.py:514]: no. the solution is: just don't jump to the end step but rather the continue step for the end step

File: rasa/core/policies/flow_policy.py [hunk 18]
```diff
@@ -470,53 +478,61 @@ def _select_next_action(
...
+                        "flow.execution.loop", previous_step_id=previous_step.id
+                    )
+                    current_step = self._select_next_step(
+                        tracker, previous_step, current_flow
+                    )
 
-            current_step = self._select_next_step(tracker, previous_step, current_flow)
+                    if current_step:
+                        # this can't be an else, because the previous if might change
+                        # this to "not None"
```
Comment (twerkmeister) [inline rasa/core/policies/flow_policy.py:517]: this comment is outdated, there are now other steps in between this if and the previous one and so an else wouldn't even be considered here
